src/rpc/bootstrap_daemon.h

```diff
@@ -0,0 +1,184 @@
+#pragma  once
+
+#include <iterator>
+#include <list>
+#include <stdexcept>
+#include <vector>
+
+#include <boost/utility/string_ref.hpp>
+
+#include "crypto/crypto.h"
+#include "cryptonote_core/cryptonote_core.h"
+#include "cryptonote_protocol/cryptonote_protocol_handler.h"
+#include "misc_log_ex.h"
+#include "net/http_client.h"
+#include "p2p/net_node.h"
+#include "p2p/p2p_protocol_defs.h"
+#include "storages/http_abstract_invoke.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "daemon.bootstrap_daemon"
+
+namespace cryptonote
+{
+
+  class bootstrap_daemon
+  {
+  public:
+    bootstrap_daemon(nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core>> *p2p)
+      : m_last_result(false)
+      , m_p2p(p2p)
+    {
+    }
+
+    bootstrap_daemon(const std::string &address, const boost::optional<epee::net_utils::http::login> &credentials)
+      : bootstrap_daemon(nullptr)
+    {
+      if (!set_server(address, credentials))
+      {
+        throw std::runtime_error("invalid bootstrap daemon address or credentials");
+      }
+    }
+
+    const std::string &address() const noexcept
+    {
+      return m_address;
+    }
+
+    bool get_height(uint64_t *height) noexcept
```

Comment:
This looks like a decent candidate for `boost::optional<uint64_t>` return value.